### PR TITLE
Move Travis CI to GitHub Actions

### PR DESCRIPTION
**Pull request by phschiele**
Hi @SteveDiamond and @rileyjmurray,

as discussed, here is an initial PR to move over the Travis CI to GitHub Actions.
You mentioned that the scope of the CI should match Travis 100%, so I stayed close to the existing implementation and reused a large portion of the existing bash scripts. I'll leave some further comments below.

**Potential follow-ups** (would be great to get your opinion on):
- Adding a cache for the dependency installation
- Further simplify e.g. the deploy step using existing marketplace actions
- Move over windows CI as well
- Make the CI more robust using an advanced dependency manager (e.g. poetry or pipenv)

**Open questions**:
- ~There is still another Travis file for cvxcore. Is that still relevant?~ Removed

**Remaining ToDos:**
- [x] Activate tests again, I disabled them to move on with development here, as they are also failing in master (related to ECOS).
- [x] Only run deploy step for pushes to master
- [x] Add relevant secrets to the GitHub repo

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    env:
+      RUNNER_OS: ${{ matrix.os }}
+      PYTHON_VERSION: ${{ matrix.python-version }}
+      COVERAGE: "False"
+      DEPLOY_PYPI_SOURCE: "${{ matrix.deploy_pypi_source == 'True' && 'True' || 'False' }}"
+      USE_OPENMP: "${{ matrix.openmp == 'True' && 'True' || 'False' }}"
+      PYPI_SERVER: ${{ secrets.PYPI_SERVER }}
+      PYPI_USER: ${{ secrets.PYPI_USER }}
+      PYPI_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
+      PYPI_API_ENDPOINT: ${{ secrets.PYPI_API_ENDPOINT }}
```
Comment (phschiele) [inline .github/workflows/build.yml:47]: These secrets need to be defined in the repo settings.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        run: |
+          if [ "$RUNNER_OS" = "macOS" ]; then CC=gcc ; fi
+          if [ "$RUNNER_OS" = "macOS" ]; then CXX=g++; fi
+          source continuous_integration/GitHub/install_dependencies.sh
+      - name: test
+        run: |
+          flake8 .
+          bash continuous_integration/GitHub/test_script.sh
+      - name: coverage
+        run: if [[ "$COVERAGE" == "True" ]]; then coveralls; fi
```
Comment (phschiele) [inline .github/workflows/build.yml:66]: It seems that the coverage run is always `False`, but I added it to keep the possibility to enable it again.

File: continuous_integration/GitHub/install_dependencies.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+# This script is meant to be called by the "install" step defined in
```
Comment (phschiele) [inline continuous_integration/GitHub/install_dependencies.sh:2]: This script was simplified quite a bit by using the existing marketplace action for setting up conda.

File: cvxpy/problems/problem.py [hunk 4]
```diff
@@ -472,7 +472,8 @@ def register_solve(cls, name, func) -> None:
         """
         cls.REGISTERED_SOLVE_METHODS[name] = func
 
-    def get_problem_data(self, solver, gp: bool = False, enforce_dpp: bool = False, verbose: bool = False):
+    def get_problem_data(self, solver, gp: bool = False, enforce_dpp: bool = False,
+                         verbose: bool = False):
```
Comment (phschiele) [inline cvxpy/problems/problem.py:476]: Just changed this line to make flake8 pass.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+name: build
+
+on:
+    pull_request:
+    push:
+        branches:
+            - master
+
+jobs:
+  cleanup-runs:
```
Comment (phschiele) [inline .github/workflows/build.yml:10]: Cancels previous GH actions workflows in your branch (if you commit again before the last tests finished).
Not executed for tag or master pushes.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    steps:
+    - uses: rokroskar/workflow-run-cleanup-action@master
+      env:
+        GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
+    if: "!startsWith(github.ref, 'refs/tags/') && github.ref != 'refs/heads/master'"
+  build:
+    runs-on: ${{ matrix.os }}
+    defaults:
+      run:
+        shell: bash -l {0}
```
Comment (phschiele) [inline .github/workflows/build.yml:21]: Required to have the conda environment activated in all bash scripts, see [use a default shell](https://github.com/conda-incubator/setup-miniconda#use-a-default-shell).

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+          if [ "$RUNNER_OS" = "macOS" ]; then CXX=g++; fi
+          source continuous_integration/GitHub/install_dependencies.sh
+      - name: test
+        run: |
+          flake8 .
+          bash continuous_integration/GitHub/test_script.sh
+      - name: coverage
+        run: if [[ "$COVERAGE" == "True" ]]; then coveralls; fi
+      - name: deploy
+        run: bash continuous_integration/GitHub/deploy_script.sh
```
Comment (phschiele) [inline .github/workflows/build.yml:68]: If the deployment to takes place is currently based on `cvxpy.__version__` in the `__init__.py` file. So I assume one should only run this stage on master/tag pushes.

Comment (SteveDiamond) [reply .github/workflows/build.yml:68]: That's correct.

File: .github/workflows/build.yml (hunk 2 above)
Comment (SteveDiamond) [reply .github/workflows/build.yml:66]: Thanks, yeah we probably want coverage again at some point.

**Comment by SteveDiamond** [comment]
Thanks so much for doing this @phschiele! You can ignore the travis.yml for cvxcore (actually delete it). To enable the tests, just mark the failing tests as skipped. They're for ECOS BB, which has known bugs.

**Comment by SteveDiamond** [comment]
Another nice thing to add would be CI going back to Python 3.5 (since in theory we support 3.5+).

File: .github/workflows/build.yml (hunk 7 above)
Comment (phschiele) [reply .github/workflows/build.yml]: Done in 8208cb5

**Comment by phschiele** [comment]
@SteveDiamond Thanks for your swift feedback, all mentioned comments are included now.

**Comment by SteveDiamond** [comment]
Ok this generally looks great, I added a few secrets but need time/help to get all of them.

**Comment by SteveDiamond** [comment]
@phschiele I believe I added all the secrets. Let me know if I missed one.

**Comment by phschiele** [comment]
> @phschiele I believe I added all the secrets. Let me know if I missed one.

@SteveDiamond Great, we can test it by reverting the last commit (and reverting again if that test works). Should we give it a try?

**Review by rileyjmurray — APPROVED** [review]
Looks good to me! Thanks for all this work, @phschiele!

**Comment by SteveDiamond** [comment]
I checked with Riley that all the secrets are there. I'm going to merge this in. Thanks again @phschiele !

**Pull request merged**